lib/Sema/CSApply.cpp

```diff
@@ -3988,6 +3988,20 @@ namespace {
       if (E->getType()->hasError())
         return E;
 
+      // If a component is already resolved, then all of them should be
+      // resolved, and we can let the expression be. This might happen when
+      // re-checking a failed system for diagnostics.
+      if (!E->getComponents().empty()
+          && E->getComponents().front().isResolved()) {
+        assert([&]{
+          for (auto &c : E->getComponents())
+            if (!c.isResolved())
+              return false;
+          return true;
+        }());
```

Comment:
I did try that, but the lambda noise felt like it outweighed any benefit over Just Writing A For Loop.